### ISSUE TITLE
Automatic editor configuration for code style

### ISSUE THREAD

**Issue by sklirg** [dotKom/onlineweb4#1477]
## Our Code Style Guidelines

We currently enforce these code styles (according to [editorconfig](https://github.com/dotKom/onlineweb4/blob/develop/.editorconfig) and [setup.cfg](https://github.com/dotKom/onlineweb4/blob/develop/setup.cfg)):
### Python
- [isort](https://github.com/timothycrosley/isort)
- [PEP8](https://www.python.org/dev/peps/pep-0008/)
- Spaces (not tabs) for indentation
- `lf` as EOL (End of Line, newlines)
- No trailing whitespace
- Blank line at end of file
- Max line length of 120 chars
- Max [Cyclomatic Complexity (McCabe)](https://en.wikipedia.org/wiki/Cyclomatic_complexity) of 10 
### Javascript
- ?
### HTML
- ?

This means that whenever we run tests, flake8 and isort will pick up any errors. But why wait for our CI to run those tests? And why should we run our full test suite locally just to find out if we pass flake8 or isort for that quick "Removes debug print" fix?

As seen recently, after enforcing both PEP8 and isort using `setup.cfg`, we've gotten a couple of commits where all they do is "Follow isort guidelines" or "Fix flake8 error". Ideally, this shouldn't need to happen, so let's set up our editors to do these checks on the fly, and warn us if something isn't right.
## Editor configuration

Let's set up the editors to follow code style guidelines we've set.

Editors and needed configuration:
### [PyCharm](https://www.jetbrains.com/pycharm/) (free w/GitHub educational)
- [x] Enforce editorconfig (on by default)
- [x] Enable PEP8 (on by default)
- [x] Enable isort ([guide](https://github.com/timothycrosley/isort/issues/258))
  - [ ] Automated setup
- [x] `lf` as EOL (Select "Unix (\n)" as described [here](https://www.jetbrains.com/pycharm/help/configuring-line-separators.html))
  - [ ] Automated setup
### Sublime Text (todo)
- [x] Enforce editorconfig (requires [plugin](https://github.com/sindresorhus/editorconfig-sublime))
- [x] Enable PEP8 (requires [plugin](https://github.com/sindresorhus/editorconfig-sublime))
- [ ] Enable isort
- [ ] `lf` as EOL
### Vim (todo)
- [ ] Enforce editorconfig
- [ ] Enable PEP8
- [ ] Enable isort
- [ ] `lf` as EOL

---

We should also update our [GitHub Wiki article](https://github.com/dotKom/onlineweb4/wiki/Konvensjoner-og-workflow-i-OW4) with these changes when we have setup for all editors we wish to support.

As always, comments, suggestions and thoughts are welcome.

TL;DR: Apply automated checks _locally_ (during development) for codestyles we enforce.

**Comment by myth** [comment]
Excellent initiative. Good stuff.

**Comment by myth** [comment]
As an addition, I propose in that we should embrace and enforce [StandardJS]( http://www.standardjs.com), which has great community traction (npm uses it etc). It also supports almost every IDE and Editor out there with plugins/templating. It has a short, clean and consise ruleset as well.

For all that its worth, we really really need this. Our JS is all over the place.

**Comment by sklirg** [comment]
I looked into this earlier this year, and I really like the [airbnb](https://github.com/airbnb/javascript) config. Have you looked at that? What do you think? 

It's pretty simple to extend the ruleset no matter which one we go for. Webkom runs their [own](https://github.com/webkom/eslint-config-webkom/blob/master/base.js) extension of airbnb.

I personally disagree with some of the StandardJS rules, but I agree we should implement some ruleset as soon as possible.

**Comment by myth** [comment]
What specifically do you disagree with? The whole point of going for something like StandardJS (which a lot of packages follow) is that is is a *standard*. Having merely guidelines which different teams fork out and make do their own thing with is just silly.

And yes, I have looked into both AirBnb guide, Google's guide and other guides such as jQuery etc.
The main problem I have with it as that they are advertised as guides and not at least an attempt to enforce it as a standard with no excuses and no exceptions.

**Comment by sklirg** [comment]
I disagree with "(Add a space before a function declaration's parentheses)[http://standardjs.com/rules.html#javascript-standard-style]" (can't link directly to the rule), I prefer [this](https://github.com/airbnb/javascript#whitespace--around-keywords). 
```
// this
function name() { … }

// not this
function name () { … }
```

Also I'd like to have curly braces for _all_ if/else statements, not just multi-line.

However, that's personal preference.

I agree. I suggested airbnb's style guide as _our guidelines_. They're widely used as well (according to npm downloads, more widely used than standard - but that's just metrics).

The problem, though, is that JavaScript doesn't _have_ a standard. This style guide is [named](http://standardjs.com/index.html#but-this-isnt-a-real-web-standard) "standard". Airbnb ships their style guide which they think is fine and enforces the styles they find makes code easier to read. And that's exactly what we want. Code quality. Whether it's inspired by airbnb, google, feross or facebook.

Regarding "disagreeing with a rule": That's the whole point of selecting a style guide. Whatever rules are present, they'll be binary; accepted or denied by CI. We can set the rules, follow the rules set by someone else, or mix it up. We can't follow the rules set by ECMA; they don't exist.

Whether we enforce double or single quotes, spaces between function names and parens, no-unused-vars, that's up to us, as _there is no one standard_. No matter which styleguide we select, 95% of the other code bases in the world will run something else (or nothing at all...).

JavaScript doesn't have a standard, and that's the reason why people have different preferences after having used the language for years and years, together with their other programming experience.

I think we should enforce a style guide. I really don't care which.

**Comment by sklirg** [comment]
Though, we have to keep in mind, when we enforce some JS guidelines, we have to fix all the failing cases before merging the config and enabling the CI tests, unless we want to keep having tests failing like during PEP8 enforcing (or possibly set up CI to pass JS style tests as long as the metrics are less than or equal the previous run). I know this is a touchy topic, and we have different views on this, but I find merging _failing branches_ to be bad practice (no matter what the cause of the failure).

**Comment by myth** [comment]
Your arguments are exactly why I argue that we should follow the movement that is actually attempting to enforce a standard.

And yes, the refactoring will be a pain. Thank god for search and replace + auto indenting.
But, we're all very happy now that we're not at 2k PEP8 errors anymore. The codebase is gradually becoming better. We need to do something about it sooner or later.

But at least this time we can decide for ourselves when we choose to activate the linter/module on our CI system. Which for the sake of sanity should be when we're at least a fair bit on the way. I still have not-so-fond memories of the night of `flake8 apps`

**Comment by sklirg** [comment]
Fair enough.

Yes, my point exactly - have a WIP-PR (or no branch at all..) which enables the linter, and merge it when it passes the JS linting. 
TeamCity has support for ESLint, so as long as we run that version of the style guide, it'll be easy peasy to set up. We can set it up whenever, and just tell TC to _not_ report to GitHub about build status for _that specific build_, and rather enable it later on when "we're ready".

Anyways, sweet. Wanna have a look at configuring PyCharm++ for this? Is it as simple as any ESLint config (ESLint & config installed and available on path in the project, and set it up in PyCharm code quality?)

Coolio.

**Comment by myth** [comment]
You have several options. There are plugins for sublime, vim, atom, emacs, brackets, VS and the IntelliJ suite, some of which perform auto-correction on the fly. PyCharm ofc supports ESLint as well, so any config file can just be dropped in. For testing purposes you could also just use npm, since we depend on node for lessc anyway. In the latter case it would just be to call `standard` (no config needed) in the root dir.

So regardless of what the decision lands on, TC and PyCharm (as well as the other major editors) support what we need.

**Comment by sklirg** [comment]
I meant a guide for dotKom :) I already have PyCharm set up for ESLint
myself.

**Comment by myth** [comment]
I didn't mean "you" as in you perticularly, but generally.

**Comment by myth** [comment]
But sure. I haven't found a way to embed automatic setup of external tools with pre-run hook tho.
The setup is very minimal, but it would have been nice to just be able to provide it in a configbundle with an installscript or something. This applies specifically to stuff like having isort (in non `-c` mode) run whenever runserver respawns. So if you know how to do that, please let me know.

**Comment by sklirg** [comment]
I think a simple install guide in the github wiki will suffice for now,
there are so many different requirements that fully automating it will be
complex.

**Comment by nornes** [comment]
+1 vote for StandardJS

**Comment by OptimusCrime** [comment]
CC på 10? war

**Comment by sklirg** [comment]
I don't think we can do much with regards to setting editor configurations automatically. :/ If someone has a great idea for how, reopen this.

**Comment by OptimusCrime** [comment]
We do this at work for IntelliJ for Java and Kotlin. It works, but might be troublesome because the config files (assuming you mean the XML project files in the .idea directory), might be different between PyCharm versions. The only reason we do this is because there are really no good standards for these languages. For Python you already have PEP8 which is well implemented in the IDE. For JavaScript you have Prettier, which you can enable post commit or something. This is much better than checking in the .idea files in my opinion.

Nice bump by the way.

**Comment by myth** [comment]
And my axe
